types/bluebird/index.d.ts

```diff
@@ -47,185 +47,237 @@ declare class Bluebird<R> implements PromiseLike<R>, Bluebird.Inspection<R> {
    */
   // Based on PromiseLike.then, but returns a Bluebird instance.
   then<U>(onFulfill?: (value: R) => U | PromiseLike<U>, onReject?: (error: any) => U | PromiseLike<U>): Bluebird<U>; // For simpler signature help.
-  then<TResult1 = R, TResult2 = never>(onfulfilled?: ((value: R) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): Bluebird<TResult1 | TResult2>;
+  then<TResult1 = R, TResult2 = never>(onfulfilled?: ((value: R) => TResult1 | PromiseLike<TResult1>) | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | null): Bluebird<TResult1 | TResult2>;
 
   /**
    * This is a catch-all exception handler, shortcut for calling `.then(null, handler)` on this promise. Any exception happening in a `.then`-chain will propagate to nearest `.catch` handler.
    *
    * Alias `.caught();` for compatibility with earlier ECMAScript version.
    */
-  catch(onReject?: (error: any) => R | PromiseLike<R> | void | PromiseLike<void>): Bluebird<R>;
-  caught(onReject?: (error: any) => R | PromiseLike<R> | void | PromiseLike<void>): Bluebird<R>;
+  catch(onReject?: (error: any) => R | PromiseLike<R | void> | undefined): Bluebird<R>;
   catch<U>(onReject?: (error: any) => U | PromiseLike<U>): Bluebird<U | R>;
-  caught<U>(onReject?: (error: any) => U | PromiseLike<U>): Bluebird<U | R>;
 
   /**
    * This extends `.catch` to work more like catch-clauses in languages like Java or C#. Instead of manually checking `instanceof` or `.name === "SomeError"`, you may specify a number of error constructors which are eligible for this catch handler. The catch handler that is first met that has eligible constructors specified, is the one that will be called.
    *
    * This method also supports predicate-based filters. If you pass a predicate function instead of an error constructor, the predicate will receive the error as an argument. The return result of the predicate will be used determine whether the error handler should be called.
    *
    * Alias `.caught();` for compatibility with earlier ECMAScript version.
-   *
-   * TODO: disallow non-objects
    */
   catch<E1 extends Error, E2 extends Error, E3 extends Error, E4 extends Error, E5 extends Error>(
     filter1: (new (...args: any[]) => E1),
     filter2: (new (...args: any[]) => E2),
     filter3: (new (...args: any[]) => E3),
     filter4: (new (...args: any[]) => E4),
     filter5: (new (...args: any[]) => E5),
-    onReject: (error: E1 | E2 | E3 | E4 | E5) => R | PromiseLike<R> | void | PromiseLike<void>,
+    onReject: (error: E1 | E2 | E3 | E4 | E5) => R | PromiseLike<R | void> | undefined,
```

Comment:
Same here: `PromiseLike<R | void>`? You sure this is type safe?
If you do want to add the `| undefined` to this function I'd change `onReject` to be an optional parameter instead. IMHO We shouldn't make the callback optional though (i.e. just remove the `| undefined`), because it would probably be a programmer error anyways if anyone forgot to pass a callback, even if the typings would be more correct with optionality.

(This comment also applies to the other changes below.)

Reply:
Also, it is supposed to throw at least an exception here as well. In this case, however, you do not need to explicitly return never.
https://github.com/petkaantonov/bluebird/blob/v3.5.1/test/mocha/catch_filter.js#L162
Review comments were very helpful. I will try to type a little more type safe.